### ISSUE TITLE
[Question] Load model with supporting structure but different size

### ISSUE THREAD

**Issue by xinranhe** [TransformerLensOrg/TransformerLens#929]
### Question

I have trained a GPT-2 structure model that is much smaller than gpt2 defined in official model. I would like to load it as HookTransformer and reuse the code to map the weight matrices. I am currently doing it in a hacky way by overwriting the HookedTransformerConfig cfg in get_pretrained_model_config with the HF config cfg that I got from my model. Just wonder is there any better way to achieve this without need to change code? Or if this is a common need, could it be supported? Thanks!

**Comment by ellenling** [comment]
@xinranhe Were you able to do this properly? I keep getting mismatches between TL and HF keys when trying to load a 2-layer / condensed gpt2 style model

Running non-strict load_state_dict to capture missing/unexpected...
 missing (expected by model but not in sd): ['embed.W_E', 'pos_embed.W_pos', 'blocks.0.ln1.w', 'blocks.0.ln1.b', 'blocks.0.ln2.w', 'blocks.0.ln2.b', 'blocks.0.attn.W_Q', 'blocks.0.attn.W_O', 'blocks.0.attn.b_Q', 'blocks.0.attn.b_O', 'blocks.0.attn.W_K', 'blocks.0.attn.W_V', 'blocks.0.attn.b_K', 'blocks.0.attn.b_V', 'blocks.0.attn.mask', 'blocks.0.attn.IGNORE', 'blocks.0.mlp.W_in', 'blocks.0.mlp.b_in', 'blocks.0.mlp.W_out', 'blocks.0.mlp.b_out', 'blocks.1.ln1.w', 'blocks.1.ln1.b', 'blocks.1.ln2.w', 'blocks.1.ln2.b', 'blocks.1.attn.W_Q', 'blocks.1.attn.W_O', 'blocks.1.attn.b_Q', 'blocks.1.attn.b_O', 'blocks.1.attn.W_K', 'blocks.1.attn.W_V', 'blocks.1.attn.b_K', 'blocks.1.attn.b_V', 'blocks.1.attn.mask', 'blocks.1.attn.IGNORE', 'blocks.1.mlp.W_in', 'blocks.1.mlp.b_in', 'blocks.1.mlp.W_out', 'blocks.1.mlp.b_out', 'ln_final.w', 'ln_final.b', 'unembed.W_U', 'unembed.b_U']
 unexpected (in sd but not expected by model): []

Checking which TL keys in the model were filled from template vs come from sd_clean:
  TL key not present in sd: embed.W_E
  TL key not present in sd: pos_embed.W_pos
  TL key not present in sd: blocks.0.ln1.w
  TL key not present in sd: blocks.0.ln1.b
  TL key not present in sd: blocks.0.ln2.w
  TL key not present in sd: blocks.0.ln2.b
  TL key not present in sd: blocks.0.attn.W_Q
  TL key not present in sd: blocks.0.attn.W_O
  TL key not present in sd: blocks.0.attn.b_Q
  TL key not present in sd: blocks.0.attn.b_O
  TL key not present in sd: blocks.0.attn.W_K
  TL key not present in sd: blocks.0.attn.W_V
  TL key not present in sd: blocks.0.attn.b_K
  TL key not present in sd: blocks.0.attn.b_V
  TL key not present in sd: blocks.0.attn.mask
  TL key not present in sd: blocks.0.attn.IGNORE
  TL key not present in sd: blocks.0.mlp.W_in
  TL key not present in sd: blocks.0.mlp.b_in
  TL key not present in sd: blocks.0.mlp.W_out
  TL key not present in sd: blocks.0.mlp.b_out
  TL key not present in sd: blocks.1.ln1.w
  TL key not present in sd: blocks.1.ln1.b
  TL key not present in sd: blocks.1.ln2.w
  TL key not present in sd: blocks.1.ln2.b
  TL key not present in sd: blocks.1.attn.W_Q
  TL key not present in sd: blocks.1.attn.W_O
  TL key not present in sd: blocks.1.attn.b_Q
  TL key not present in sd: blocks.1.attn.b_O
  TL key not present in sd: blocks.1.attn.W_K
  TL key not present in sd: blocks.1.attn.W_V
  TL key not present in sd: blocks.1.attn.b_K
  TL key not present in sd: blocks.1.attn.b_V
  TL key not present in sd: blocks.1.attn.mask
  TL key not present in sd: blocks.1.attn.IGNORE
  TL key not present in sd: blocks.1.mlp.W_in
  TL key not present in sd: blocks.1.mlp.b_in
  TL key not present in sd: blocks.1.mlp.W_out
  TL key not present in sd: blocks.1.mlp.b_out
  TL key not present in sd: ln_final.w
  TL key not present in sd: ln_final.b
  TL key not present in sd: unembed.W_U
  TL key not present in sd: unembed.b_U
  -> template-copied count: 42
  -> differing count: 0

For differing TL keys, best HF candidate (showing up to best 3 candidates):

HF keys present but not matched to any TL candidate under the limited candidate set tried above:
 Total HF keys: 29
 Unused HF keys (count): 29
 Sample unused HF keys (first 80):
  transformer.wte.weight
  transformer.wpe.weight
  transformer.h.0.ln_1.weight
  transformer.h.0.ln_1.bias
  transformer.h.0.attn.c_attn.weight
  transformer.h.0.attn.c_attn.bias
  transformer.h.0.attn.c_proj.weight
  transformer.h.0.attn.c_proj.bias
  transformer.h.0.ln_2.weight
  transformer.h.0.ln_2.bias
  transformer.h.0.mlp.c_fc.weight
  transformer.h.0.mlp.c_fc.bias
  transformer.h.0.mlp.c_proj.weight
  transformer.h.0.mlp.c_proj.bias
  transformer.h.1.ln_1.weight
  transformer.h.1.ln_1.bias
  transformer.h.1.attn.c_attn.weight
  transformer.h.1.attn.c_attn.bias
  transformer.h.1.attn.c_proj.weight
  transformer.h.1.attn.c_proj.bias
  transformer.h.1.ln_2.weight
  transformer.h.1.ln_2.bias
  transformer.h.1.mlp.c_fc.weight
  transformer.h.1.mlp.c_fc.bias
  transformer.h.1.mlp.c_proj.weight
  transformer.h.1.mlp.c_proj.bias
  transformer.ln_f.weight
  transformer.ln_f.bias
  lm_head.weight

**Comment by jlarson4** [comment]
Hi @xinranhe  & @satyrmipt ! The HookedTransformer.from_pretrained path you're patching is being phased out in favor of TransformerLens 3.0's TransformerBridge.boot_transformers. The bridge accepts local directory paths directly and supports local models via the hf_model= parameter. I'd recommend [migrating to TransformerLens 3.0](https://transformerlensorg.github.io/TransformerLens/content/migrating_to_v3.html), and trying the following:

```
from transformer_lens.model_bridge import TransformerBridge

# Option A: pass a local path
bridge = TransformerBridge.boot_transformers("/path/to/local/gpt2")

# Option B: pre-load and pass the model
from transformers import AutoModelForCausalLM, AutoTokenizer
hf_model = AutoModelForCausalLM.from_pretrained("/path/to/local/gpt2")
tokenizer = AutoTokenizer.from_pretrained("/path/to/local/gpt2")
bridge = TransformerBridge.boot_transformers(
    "/path/to/local/gpt2",
    hf_model=hf_model,
    tokenizer=tokenizer,
)
```

Closing this in favor of the bridge path. Let me know if you run into any issues migrating to TransformerBridge or if your issue continues to persist, I would be happy to help!
